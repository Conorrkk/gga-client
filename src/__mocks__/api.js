const teamMock = {
    teamName: 'Bredagh',
    teamLevel: 'Senior',
}

export const apiRequests = () => {
    return {
        getTotalGoals: jest.fn(() => 2),
        getTotalPoints: jest.fn(() => 10),
        getTeamById: jest.fn(() => {
            return teamMock;
        })
    };
}