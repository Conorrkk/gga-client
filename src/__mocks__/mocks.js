export const player1 = {
    _id: 1,
    playerId: 1,
    playerName: 'Conor Kelly',
    stats: {
        goal_from_play: 0,
        goal_from_dead: 1,
        point_from_play: 2,
        point_from_dead: 0,
        wide: 0
    }
}

export const player2 = {
    _id: 2,
    playerId: 2,
    playerName: 'Harry Potter',
    stats: {
        goal_from_play: 0,
        goal_from_dead: 1,
        point_from_play: 2,
        point_from_dead: 0,
        wide: 2
    }
}

export const playersMock = [player1, player2]; 

export const match = {
    _id: "123",
    teams: {
        teamId: "123",
        oppositionTeam: "Galway",
        players: [player1, player2]
    },
    matchDate: 8/30/2023,
    goalAgainst: 1,
    pointAgainst: 14
};

export const matches = [match];