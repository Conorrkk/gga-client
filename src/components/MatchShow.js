import { useState, useEffect } from "react";
import { Button, Col, Card, Row } from "react-bootstrap";
import { getTeamById, getTotalGoals, getTotalPoints } from "../api";
import { useNavigate } from "react-router-dom";
import { FaTrashAlt, FaEye } from "react-icons/fa";
import "../styles.css";

function MatchShow({ match, onDelete }) {
  // the team in this match
  const [team, setTeam] = useState("");
  // users team scores in this match
  const [userGoals, setUserGoals] = useState(0);
  const [userPoints, setUserPoints] = useState(0);
  // opposing team's name
  const oppositionName = match.teams.oppositionTeam;

  // creating a new date using the date from db and reformatting it from iso
  const date = new Date(match.matchDate);
  const options = { day: "numeric", month: "numeric", year: "numeric" };
  const configuredDate = date.toLocaleDateString(undefined, options);

  // for page nav
  const navigate = useNavigate();

  // gets the total goals for the user's team
  useEffect(() => {
    const getGoals = async () => {
      const goalScored = await getTotalGoals(match._id);
      setUserGoals(goalScored.totalGoals);
    };
    getGoals();
  }, [match._id]);

  // gets the total points for the user's team
  useEffect(() => {
    const getPoints = async () => {
      const pointScored = await getTotalPoints(match._id);
      setUserPoints(pointScored.totalPoints);
    };
    getPoints();
  }, [match]);

  // deletes a match
  const handleDelete = () => {
    onDelete(match._id);
  };

  // when user clicks on a match card go to match overview
  const handleClick = () => {
    const matchId = match._id;
    navigate(`/matchOverview/${matchId}`);
  };

  // loads user's team name
  useEffect(() => {
    const getTeam = async () => {
      try {
        const teamId = match.teams.teamId;
        const response = await getTeamById(teamId);
        setTeam(response);
      } catch (error) {
        console.error("Error getting team name:", error);
      }
    };
    getTeam();
  }, [match]);

  return (
    <Col
      sm={{ span: 8 }}
      md={{ span: 8, offset: 2 }}
      lg={{ span: 8, offset: 2 }}
    >
      <Card className="styled-card mx-4 my-4 item-hover">
        <Card.Header>{configuredDate}</Card.Header>
        <Card.Body >
          <Card.Title>
            {team && team.teamName} {team && team.teamLevel} vs {oppositionName}
          </Card.Title>
          <Card.Text>
            {userGoals}-{userPoints} : {match.goalAgainst}-{match.pointAgainst}
          </Card.Text>
          <Row>
            <Col className="d-flex justify-content-start">
              <Button
                variant="outline-dark"
                onClick={handleClick}
                className="styled-icon-button mx-1"
              >
                <FaEye />
              </Button>
            </Col>
            <Col className="d-flex justify-content-end">
              <Button
                variant="outline-danger"
                onClick={handleDelete}
                className="styled-icon-button mx-1"
              >
                <FaTrashAlt style={{ color: "red" }}></FaTrashAlt>
              </Button>
            </Col>
          </Row>
        </Card.Body>
      </Card>
    </Col>
  );
}

export default MatchShow;
