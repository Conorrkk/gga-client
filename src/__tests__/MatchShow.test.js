import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { match } from '../__mocks__/mocks';
import { BrowserRouter, useNavigate } from 'react-router-dom';
import MatchShow from '../components/MatchShow';
import { render, fireEvent, waitFor } from '@testing-library/react';

const teamMock = {
    teamName: 'Bredagh',
    teamLevel: 'Senior',
}

jest.mock('../api', () => {
    return {
        getTotalGoals: jest.fn(() => 2),
        getTotalPoints: jest.fn(() => 10),
        getTeamById: jest.fn(() => {
            return teamMock;
        })
    };
});

const handleDeleteMock = jest.fn();
const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom')),
    useNavigate: () => mockUseNavigate,
}));

const setState = jest.fn();
jest.spyOn(React, 'useState').mockImplementation(initState => [initState, setState]);

it('renders correctly', async () => {
    let tree = renderer
            .create(<BrowserRouter><MatchShow match={match} onDelete={() => {return;}}/></BrowserRouter>)
            .toJSON();

    await waitFor(() => {
        expect(tree).toMatchSnapshot();
    });
});

it('calls onDelete with the matchId whenever the delete button is clicked', async () => {
    const { getByLabelText } = render(<BrowserRouter><MatchShow match={match} onDelete={handleDeleteMock} /></BrowserRouter>);

    const deleteButton = getByLabelText("Delete");

    await fireEvent.click(deleteButton);

    await waitFor(() => {
        expect(handleDeleteMock).toHaveBeenCalledWith("123");
    });
});

it('navigates to match overview once the view button is clicked', async () => {
    const { getByLabelText } = await render(<BrowserRouter><MatchShow match={match} onDelete={handleDeleteMock} /></BrowserRouter>);

    const viewButton = getByLabelText("ViewMatch");

    await fireEvent.click(viewButton);

    await waitFor(() => {
        expect(mockUseNavigate).toHaveBeenCalledWith('/matchOverview/123');
    });

});