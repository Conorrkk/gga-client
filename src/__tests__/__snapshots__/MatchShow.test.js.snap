// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="col-lg-8 col-md-8 col-sm-8 offset-lg-2 offset-md-2"
>
  <div
    className="styled-card mx-4 my-4 card"
  >
    <div
      className="card-header"
    >
      1/1/1970
    </div>
    <div
      className="card-body"
    >
      <div
        aria-label="card-title"
        className="card-title h5"
      >
         
         vs 
        Galway
      </div>
      <p
        className="card-text"
      >
        0
        -
        0
         : 
        1
        -
        14
      </p>
      <div
        className="row"
      >
        <div
          className="d-flex justify-content-start col"
        >
          <button
            aria-label="ViewMatch"
            className="styled-icon-button mx-1 btn btn-outline-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
              />
            </svg>
          </button>
        </div>
        <div
          className="d-flex justify-content-end col"
        >
          <button
            aria-label="Delete"
            className="styled-icon-button mx-1 btn btn-outline-danger"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": "red",
                }
              }
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
